### PR TITLE
all: Make external channels work with a UrlRoot

### PR DESCRIPTION
**Pull request by mvollmer**
When a UrlRoot was in use, almost all URLs for external channels would
forget to include it.

The effect was that diagnostic reports can't be downloaded,
applications don't have icons, and virtual machine consoles don't
work.

### CODE REVIEW

**Comment by mvollmer** [comment]
This uses cockpit.transport.uri in a ugly way, but I have shied away from extending cockpit.js itself since I am not sure of the rules anymore... does it make sense to extend ir or should we rather put something in pkg/lib/ that interested pages can statically import?

**Comment by mvollmer** [comment]
Would be good to have tests for this.  Also, when hopping directly to another machine, this is still broken.

**Comment by mvollmer** [comment]
> Also, when hopping directly to another machine, this is still broken.

No, it works, I had tested it wrong.

**Comment by mvollmer** [comment]
Looks like a broke the VM console...

**Comment by mvollmer** [comment]
> Looks like I broke the VM console...

Should be fixed now, there was one forward slash too many in the resulting URL.

**Comment by martinpitt** [comment]
Ah, please remove the "needswork" label when fixing stuff.. Retrying tests, they look mostly good now.

File: pkg/playground/speed.html [hunk 1]
```diff
@@ -76,7 +76,7 @@ <h1>Download Test</h1>
             <tbody>
                 <tr>
                     <td><label class="control-label" for="message">Path</label></td>
-                    <td><input class="form-control" id="download-path" value="/dev/sda"></td>
+                    <td><input class="form-control" id="download-path" value="/etc/os-release"></td>
```
Comment (martinpitt) [inline pkg/playground/speed.html]: /dev/sda wasn't ideal for sure, as normal users can't read it. But /etc/os-release doesn't provide much of a speed test. I suppose this is a debugging leftover?

Comment (mvollmer) [reply pkg/playground/speed.html]: Ah, I wasn't thinking of testing speed at all.  I revert this.

**Comment by mvollmer** [comment]
> Ah, please remove the "needswork" label when fixing stuff..

That was for adding a test. Let's see how hard that is...

**Comment by martinpitt** [comment]
A lot of tests fail with
```
  File "/build/cockpit/test/common/testlib.py", line 485, in login_and_go
    href = urlroot + path
TypeError: can only concatenate str (not "NoneType") to str
```

File: pkg/playground/speed.html (hunk 1 above)
Comment (mvollmer) [reply pkg/playground/speed.html]: Done

**Comment by mvollmer** [comment]
> href = urlroot + path

Oops, I didn't account for the `path == None` case.

**Comment by martinpitt** [comment]
The [container tests](https://209.132.184.41:8493/logs/pull-11853-20190528-074949-0ab3524c-container-kubernetes/log.html) need updating, too:

    TypeError: login_and_go() got an unexpected keyword argument 'urlroot'

**Comment by mvollmer** [comment]
> The container tests need updating, too:

They had their own definition of login_and_go.  Updated as well.

**Comment by martinpitt** [comment]
The [testWithUrlRoot check](https://209.132.184.41:8493/logs/pull-11853-20190528-193232-0753acb6-verify-rhel-7-6-distropkg/log.html#265)  fails on -distropkg, this needs to be special-cased.

**Comment by martinpitt** [comment]
Pushed trivial fix.

**Pull request merged**